nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ExtractGrok.java

```diff
@@ -209,78 +287,94 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
-                });
-                conFlowfile = session.putAttribute(conFlowfile, CoreAttributes.MIME_TYPE.key(), APPLICATION_JSON);
-                session.getProvenanceReporter().modifyContent(conFlowfile, "Replaced content with parsed Grok fields and values", stopWatch.getElapsed(TimeUnit.MILLISECONDS));
-                session.transfer(conFlowfile, REL_MATCH);
+            if (results.isEmpty()) {
+                session.transfer(flowFile, REL_NO_MATCH);
+                getLogger().debug("Did not match any Grok Expression for FlowFile {}", new Object[]{flowFile});
+                return;
+            }
+
+            String matchedExpressions = StringUtils.join(matchedExpressionList, expressionSeparator);
+            flowFile = session.putAttribute(flowFile, matchedExpressionAttribute, matchedExpressions);
 
-                break;
+            switch (context.getProperty(DESTINATION).getValue()) {
+                case FLOWFILE_ATTRIBUTE:
+                    Map<String, String> grokResults = new HashMap<>();
+                    for (Map.Entry<String, Object> entry : results.entrySet()) {
+                        if (null != entry.getValue()) {
+                            grokResults.put(resultPrefix + entry.getKey(), entry.getValue().toString());
+                        }
+                    }
+                    flowFile = session.putAllAttributes(flowFile, grokResults);
+                    session.getProvenanceReporter().modifyAttributes(flowFile);
+                    session.transfer(flowFile, REL_MATCH);
+                    getLogger().info("Matched {} Grok Expressions and added attributes to FlowFile {}", new Object[]{grokResults.size(), flowFile});
+                    break;
+                case FLOWFILE_CONTENT:
+                    final ObjectMapper objectMapper = new ObjectMapper();
+                    FlowFile conFlowfile = session.write(flowFile, (in, out) -> {
+                        out.write(objectMapper.writeValueAsBytes(results));
+                    });
+
+                    conFlowfile = session.putAttribute(conFlowfile, CoreAttributes.MIME_TYPE.key(), APPLICATION_JSON);
+                    session.getProvenanceReporter().modifyContent(conFlowfile, "Replaced content with parsed Grok fields and values", stopWatch.getElapsed(TimeUnit.MILLISECONDS));
+                    session.transfer(conFlowfile, REL_MATCH);
+                    break;
+            }
+        }catch(ProcessException t){
+            flowFile = session.putAttribute(flowFile, getClass().getSimpleName() + ".exception", t.getMessage());
+            session.transfer(flowFile, REL_NO_MATCH);
+            getLogger().error("Did not match any Grok Expression for FlowFile {}", new Object[]{flowFile});
+            return;
         }
     }
 
+    public static Validator validateMultipleFilesExist() {
+        return (subject, input, context) -> {
+            for (String s : input.split(PATTERN_FILE_LIST_SEPARATOR)) {
```

Comment:
The logic here doesn't appear to be right. This is going to create a validator for the first file that it encounters but not validate the others.

Reply:
oops.